sampo/generator/utils/graph_node_operations.py

```diff
@@ -1,28 +1,25 @@
-import queue
-
 from sampo.schemas.graph import GraphNode
 
 
-def count_node_ancestors(finish: GraphNode, root: GraphNode) -> int:
+def count_ancestors(first_ancestors: list[GraphNode], root: GraphNode) -> int:
     """
     Counts the number of ancestors of the whole graph.
 
-    :param finish: The node for which ancestors are to be counted.
+    :param first_ancestors: First ancestors of ancestors which must be counted.
     :param root: The root node of the graph.
     :return:
     """
-    q = queue.Queue()
-    count = 0
+    q = list(first_ancestors)
+    count = len(first_ancestors)
     used = set()
     used.add(root)
-    q.put(finish)
-    while not q.empty():
-        node = q.get()
+    while q:
+        node = q.pop()
         for parent in node.parents:
             if parent in used:
                 continue
             used.add(parent)
-            q.put(parent)
+            q.insert(0, parent)
```

Comment:
This operation leads to `O(n)` complexity, where `n = len(q)`.
`queue.SimpleQueue` should be used here instead of list